### ISSUE TITLE
Investigate Mono failures on Travis

### ISSUE THREAD

**Issue by rynowak** [aspnet/EventNotification#16]
Some tests (to do with proxies for IList) are failing on Travis with invalid IL exceptions.

See: b76c26aa3ce0ac66ed4bc68cda49e5f3b30b3431

**Comment by ryanbrandenburg** [comment]
After some investigating this seems to be a problem with the IL generated in ProxyTypeEmitter (specifically in AddProperties). It doesn't actually seem to be a problem with Mono since running PEVerify gives errors. Here are the PEVerify errors we get from saving the generations from that set of tests to a dll:

```
Microsoft (R) .NET Framework PE Verifier.  Version  4.0.30319.0
Copyright (c) Microsoft Corporation.  All rights reserved.

[IL]: Error: [C:\Users\rybrande\Projects\EventNotification\test\Microsoft.Extensions.DiagnosticAdapter.Test\Microsoft.Extensions.DiagnosticAdapter.ProxyAssembly.dll : Proxy_From_Person_To_IPerson_4::get_Address][offset 0x00000016][found ref 'Microsoft.Extensions.DiagnosticAdapter.DiagnosticSourceAdapterTest+IAddress'][expected ref 'Microsoft.Extensions.DiagnosticAdapter.DiagnosticSourceAdapterTest+Address'] Unexpected type on the stack.
[IL]: Error: [C:\Users\rybrande\Projects\EventNotification\test\Microsoft.Extensions.DiagnosticAdapter.Test\Microsoft.Extensions.DiagnosticAdapter.ProxyAssembly.dll : Proxy_From_Person_To_IPerson_4::get_Address][offset 0x0000001B][found ref 'System.Object'][expected ref 'Microsoft.Extensions.DiagnosticAdapter.DiagnosticSourceAdapterTest+IAddress'] Unexpected type on the stack.
[IL]: Error: [C:\Users\rybrande\Projects\EventNotification\test\Microsoft.Extensions.DiagnosticAdapter.Test\Microsoft.Extensions.DiagnosticAdapter.ProxyAssembly.dll : Proxy_From_Person_To_IPerson_6::get_Address][offset 0x00000016][found ref 'Microsoft.Extensions.DiagnosticAdapter.DiagnosticSourceAdapterTest+IAddress'][expected ref 'Microsoft.Extensions.DiagnosticAdapter.DiagnosticSourceAdapterTest+Address'] Unexpected type on the stack.
[IL]: Error: [C:\Users\rybrande\Projects\EventNotification\test\Microsoft.Extensions.DiagnosticAdapter.Test\Microsoft.Extensions.DiagnosticAdapter.ProxyAssembly.dll : Proxy_From_Person_To_IPerson_6::get_Address][offset 0x0000001B][found ref 'System.Object'][expected ref 'Microsoft.Extensions.DiagnosticAdapter.DiagnosticSourceAdapterTest+IAddress'] Unexpected type on the stack.
[IL]: Error: [C:\Users\rybrande\Projects\EventNotification\test\Microsoft.Extensions.DiagnosticAdapter.Test\Microsoft.Extensions.DiagnosticAdapter.ProxyAssembly.dll : Proxy_From_CoolPerson_To_ICoolPerson_10::get_Address][offset 0x00000016][found ref 'Microsoft.Extensions.DiagnosticAdapter.DiagnosticSourceAdapterTest+IAddress'][expected ref 'Microsoft.Extensions.DiagnosticAdapter.DiagnosticSourceAdapterTest+Address'] Unexpected type on the stack.
[IL]: Error: [C:\Users\rybrande\Projects\EventNotification\test\Microsoft.Extensions.DiagnosticAdapter.Test\Microsoft.Extensions.DiagnosticAdapter.ProxyAssembly.dll : Proxy_From_CoolPerson_To_ICoolPerson_10::get_Address][offset 0x0000001B][found ref 'System.Object'][expected ref 'Microsoft.Extensions.DiagnosticAdapter.DiagnosticSourceAdapterTest+IAddress'] Unexpected type on the stack.
[IL]: Error: [C:\Users\rybrande\Projects\EventNotification\test\Microsoft.Extensions.DiagnosticAdapter.Test\Microsoft.Extensions.DiagnosticAdapter.ProxyAssembly.dll : Proxy_From_CoolPerson_To_ICoolPerson_10::get_Coolness][offset 0x0000000C][found value 'System.Decimal'] Unexpected type on the stack.
[IL]: Error: [C:\Users\rybrande\Projects\EventNotification\test\Microsoft.Extensions.DiagnosticAdapter.Test\Microsoft.Extensions.DiagnosticAdapter.ProxyAssembly.dll : Proxy_From_HasListProperty_To_IHasReadOnlyListProperty_12::get_ListProperty][offset 0x00000016][found ref 'System.Collections.Generic.IReadOnlyList`1[System.String]'][expected ref 'System.Collections.Generic.IList`1[System.String]'] Unexpected type on the stack.
[IL]: Error: [C:\Users\rybrande\Projects\EventNotification\test\Microsoft.Extensions.DiagnosticAdapter.Test\Microsoft.Extensions.DiagnosticAdapter.ProxyAssembly.dll : Proxy_From_HasListProperty_To_IHasReadOnlyListProperty_12::get_ListProperty][offset 0x0000001B][found ref 'System.Collections.Generic.IEnumerable`1[System.String]'][expected ref 'System.Collections.Generic.IReadOnlyList`1[System.String]'] Unexpected type on the stack.
9 Error(s) Verifying Microsoft.Extensions.DiagnosticAdapter.ProxyAssembly.dll
```

So it would seem to me that Mono is more strict, whereas otherwise you wouldn't notice unless you run PEVerify. Since PEVerify seems the be the canonical "truth" of IL correctness I would say that the IL we're generating here is wrong in a way I haven't quite grasped yet but having to do with casting/type manipulation.

**Comment by rynowak** [comment]
My guess is that we need to add a `il.Emit(OpCodes.Castclass, targetType);` after 
https://github.com/aspnet/EventNotification/blob/dev/src/Microsoft.Extensions.DiagnosticAdapter/Internal/ProxyTypeEmitter.cs#L398

Give that a try

**Comment by Eilon** [comment]
@ryanbrandenburg is this `2 - Working`? :smile:

**Comment by ryanbrandenburg** [comment]
Changed the status.

@rynowak I had tried that one, when I do that I get:

```
Microsoft (R) .NET Framework PE Verifier.  Version  4.0.30319.0
Copyright (c) Microsoft Corporation.  All rights reserved.

[IL]: Error: [C:\Users\rybrande\Projects\EventNotification\test\Microsoft.Extensions.DiagnosticAdapter.Test\Microsoft.Extensions.DiagnosticAdapter.ProxyAssembly.dll : Proxy_From_Person_To_IPerson_4::get_Address][offset 0x0000001B][found ref 'System.Object'][expected ref 'Microsoft.Extensions.DiagnosticAdapter.DiagnosticSourceAdapterTest+IAddress'] Unexpected type on the stack.
[IL]: Error: [C:\Users\rybrande\Projects\EventNotification\test\Microsoft.Extensions.DiagnosticAdapter.Test\Microsoft.Extensions.DiagnosticAdapter.ProxyAssembly.dll : Proxy_From_Person_To_IPerson_6::get_Address][offset 0x0000001B][found ref 'System.Object'][expected ref 'Microsoft.Extensions.DiagnosticAdapter.DiagnosticSourceAdapterTest+IAddress'] Unexpected type on the stack.
[IL]: Error: [C:\Users\rybrande\Projects\EventNotification\test\Microsoft.Extensions.DiagnosticAdapter.Test\Microsoft.Extensions.DiagnosticAdapter.ProxyAssembly.dll : Proxy_From_CoolPerson_To_ICoolPerson_8::get_Address][offset 0x0000001B][found ref 'System.Object'][expected ref 'Microsoft.Extensions.DiagnosticAdapter.DiagnosticSourceAdapterTest+IAddress'] Unexpected type on the stack.
[IL]: Error: [C:\Users\rybrande\Projects\EventNotification\test\Microsoft.Extensions.DiagnosticAdapter.Test\Microsoft.Extensions.DiagnosticAdapter.ProxyAssembly.dll : Proxy_From_CoolPerson_To_ICoolPerson_8::get_Coolness][offset 0x0000000C][found value 'System.Decimal'] Unexpected type on the stack.
[IL]: Error: [C:\Users\rybrande\Projects\EventNotification\test\Microsoft.Extensions.DiagnosticAdapter.Test\Microsoft.Extensions.DiagnosticAdapter.ProxyAssembly.dll : Proxy_From_HasListProperty_To_IHasReadOnlyListProperty_12::get_ListProperty][offset 0x0000001B][found ref 'System.Object'][expected ref 'System.Collections.Generic.IReadOnlyList`1[System.String]'] Unexpected type on the stack.
5 Error(s) Verifying Microsoft.Extensions.DiagnosticAdapter.ProxyAssembly.dll
```

Which does seem less wrong (or at least now wrong in all the same ways), but still isn't quite there.

**Comment by ryanbrandenburg** [comment]
Fixed in #36
